### ISSUE TITLE
Should loopback and other trusted urls be upgraded?

### ISSUE THREAD

**Issue by jonathanKingston** [w3c/webappsec-upgrade-insecure-requests#14]
In https://bugzilla.mozilla.org/show_bug.cgi?id=1447784#c5 it isn't clear if loopback should be ignored. I wanted to ask here as this likely will be more visible than the comment (sorry for the duplication).

TL;DR loopback 127.0.0.0/8 is considered secure elsewhere but in this case we attempt to upgrade, both Chrome and Firefox behave in a consistent manner.

I think the headings of the spec and some of the text could be a little clearer for this edge case if the decision is that they should remain upgraded.

Also it's worth noting that this isn't covered by web platform testing.

**Comment by mikewest** [comment]
TL;DR: `Upgrade-Insecure-Requests` does not intend to upgrade `http://127.0.0.1/`.

Step 1.1 of https://w3c.github.io/webappsec-upgrade-insecure-requests/#upgrade-request excludes ["_a priori_ authenticated" URLs](https://w3c.github.io/webappsec-mixed-content/#a-priori-authenticated-url), which are defined in terms of Secure Contexts' [Is URL trustworthy?](https://www.w3.org/TR/secure-contexts/#is-url-trustworthy). If Chrome's trying to upgrade these URLs, Chrome should stop doing so.

> I think the headings of the spec and some of the text could be a little clearer for this edge case if the decision is that they should remain upgraded.

I'd be happy to accept a PR clarifying in a note that this means that things like loopback and localhost aren't upgraded. :)

> Also it's worth noting that this isn't covered by web platform testing.

I don't think it's possible to test this via WPT, as WPT does not run on `localhost` or `127.0.0.1`, but `web-platform.test`, etc. It _happens_ to run on `127.0.0.1` in some vendor's internal test architecture, but that's not something that the tests can assume.

**Comment by annevk** [comment]
Is there an issue against web-platform-tests about not being able to test this? It seems important to make this possible and at least when you locally configure web-platform-tests there would be ways to do this I think (it binds its domains to 127.0.0.1 already).

**Comment by mikewest** [comment]
@jgraham and @foolip probably have opinions. My impression is that we intentionally don't assume that WPT is running locally in order to support deployments like https://w3c-test.org/. We don't have `file:` tests for the same reason. But I'll defer to folks who know more about the infrastructure challenges than I do.

**Comment by foolip** [comment]
So, not everything in WPT is testable via w3c-test.org manually in a browser, but it is important that it's possible to use domains other than web-platform.test, possibly ones that resolve via public DNS, as we think that will come in handy for testing mobile OSes where we can't run the server on the device or perhaps even change the /etc/hosts equivalent.

That being said, I think a mechanism to testing against 127.0.0.1 would be fine if those tests are just skipped when the sever isn't local.

**Comment by jonathanKingston** [comment]
> Step 1.1 of https://w3c.github.io/webappsec-upgrade-insecure-requests/#upgrade-request excludes "a priori authenticated" URLs, which are defined in terms of Secure Contexts' Is URL trustworthy?.

So my reading of this is that:
- the algorithm should continue (it doesn't mention returning).
- this only applies to navigation requests

>   Chrome should stop doing so.
The STR in the bug mentioned was reproducible on Chrome stable for me anyway:  https://bugzilla.mozilla.org/show_bug.cgi?id=1447784

> I'd be happy to accept a PR clarifying in a note that this means that things like loopback and localhost aren't upgraded. :)

I actually think that the points should be renumbered and point 1. should be:
```
1. if both of the following criteria are met return: 
    1.1 request’s url is a priori authenticated
    1.2 request’s url’s host is a preloadable HSTS host
2. If request is a navigation request, append a header named Upgrade-Insecure-Requests with a value of 1 to request’s header-list
...
```

> I don't think it's possible to test this via WPT, as WPT does not run on localhost or 127.0.0.1, but web-platform.test, etc. It happens to run on 127.0.0.1 in some vendor's internal test architecture, but that's not something that the tests can assume.

Perhaps there should be a mechanism of treating a hostname as local for testing purposes, would that work instead?

**Comment by foolip** [comment]
> Perhaps there should be a mechanism of treating a hostname as local for testing purposes, would that work instead?

Maybe. @mikewest, do you know where the special-casing of "local" happens, and whether injecting "treat these things as local" there would be on the table? I imagine that it's a bit security sensitive, just like pretending that various origins are secure which actually are not?

**Comment by ericlaw1979** [comment]
I also found this confusing, and there are definitely differing behaviors across browsers.

To clarify the spec confusion, could we first add a line before step 3 of `4.1. Upgrade request to an a priori authenticated URL, if appropriate` mentioning that if a URL is *already* a priori authenticated, then skip the remaining steps?

I built a test page that explores how browsers currently behave here: https://fiddlerbook.com/test/csp/uirheader.aspx

In Chrome 73 Canary, I see that Chrome is *not* upgrading HTTP/localhost and HTTP/127.0.0.1 references when they appear directly within the markup. However, when these targets are specified in the `Location` header of a 302 response, the HTTP is changed to HTTPS. Within Chrome, CSPContext::ModifyRequestUrlForCsp checks `url->SchemeIs(url::kHttpScheme) && !IsOriginSecure(*url)` before changing the scheme to HTTPS. However, the URLRequestJob::IsRedirectResponse function checks only `location->SchemeIs("http")`
